lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -0,0 +1,922 @@
+    parameters[0] = angle;
+    parameters[1] = pi_2;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+    parameters[0] = negPi_2;
+    parameters[1] = zero;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// XOp decompositions
+//===----------------------------------------------------------------------===//
+
+struct CXToCZ : public OpRewritePattern<quake::XOp> {
+  using OpRewritePattern<quake::XOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("CXToCZ"); }
+
+  LogicalResult matchAndRewrite(quake::XOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!quake::isAllReferences(op))
+      return failure();
+    if (failed(checkNumControls(op, 1)))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.create<quake::ZOp>(loc, op.getControls(), target);
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct CCXToCCZ : public OpRewritePattern<quake::XOp> {
+  using OpRewritePattern<quake::XOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("CCXToCCZ"); }
+
+  LogicalResult matchAndRewrite(quake::XOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!quake::isAllReferences(op))
+      return failure();
+    if (failed(checkNumControls(op, 2)))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.create<quake::ZOp>(loc, op.getControls(), target);
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+// quake.x target
+// ───────────────────────────────
+// quake.phased_rx(π, 0) target
+struct XToPhasedRx : public OpRewritePattern<quake::XOp> {
+  using OpRewritePattern<quake::XOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("XToPhasedRx"); }
+
+  LogicalResult matchAndRewrite(quake::XOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!op.getControls().empty())
+      return failure();
+    if (!quake::isAllReferences(op))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+
+    // Necessary/Helpful constants
+    ValueRange noControls;
+    Value zero = createConstant(loc, 0.0, rewriter.getF64Type(), rewriter);
+    Value pi = createConstant(loc, M_PI, rewriter.getF64Type(), rewriter);
+
+    ValueRange parameters = {pi, zero};
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
```

Comment:
replaceOpWithNewOp? Maybe I'm wrong that it's applicable here

Reply:
We could use it, but I don't see much difference since these are operation that don't return values.